File: 04-stack/00020-Valid_Parentheses/00020-valid_parentheses.cpp
// 20: Valid Parentheses
// https://leetcode.com/problems/valid-parentheses/


#include <iostream>
#include <stack>

using namespace std;


class Solution {
public:
    // SOLUTION
    bool isValid (string s) {
        stack<char> parentheses;

        for (auto &c : s) {
            switch(c) {
                case '{': parentheses.push('}'); break;
                case '[': parentheses.push(']'); break;
                case '(': parentheses.push(')'); break;
                default:
                    if (parentheses.size()==0 || c!=parentheses.top())
                        return false;
                    else parentheses.pop();
            }
        }

        return parentheses.empty();
    }
};


int main() {
    Solution o;

    // INPUT
    string s = "()[]{}";

    // OUTPUT
    auto result = o.isValid(s);
    cout<<(result ? "true" : "false")<<endl;

    return 0;
}

File: 04-stack/00020-Valid_Parentheses/00020-valid_parentheses.go
// 20: Valid Parentheses
// https://leetcode.com/problems/valid-parentheses/

package main

import "fmt"

// SOLUTION
func isValid(s string) bool {
	var parentheses []rune

	for _, c := range s {
		switch c {
			case '{':
				parentheses = append(parentheses, '}')
			case '[':
				parentheses = append(parentheses, ']')
			case '(':
				parentheses = append(parentheses, ')')
			default:
				if len(parentheses)==0 || c!=parentheses[len(parentheses)-1] {
					return false
				} else {
					parentheses = parentheses[:len(parentheses)-1]
				}
		}
	}

	return len(parentheses)==0
}


func main() {
	// INPUT
	s := "()[]{}"

	// OUTPUT
	result := isValid(s)
	fmt.Println(result)
}